Add ServerPlayerEvents.ALLOW_DEATH event

williambl opened the pull request:
### PR Contents

This PR adds a new event, `ServerPlayerEvents.ALLOW_DEATH`. This is called when a player has just taken fatal damage. Mods can return `false` from this to cancel the death.

### Rationale

Mods may want to cancel a player's death in certain circumstances - [Gunpowder](https://github.com/Gunpowder-MC/Gunpowder/blob/c62a696eda68ca5648d78ed10779bf363efcb506/gunpowder-base/src/main/java/io/github/gunpowder/mixin/base/LivingEntityMixin_Base.java#L43) and [Haema](https://github.com/williambl/haema/blob/c7026d9ee18e9414b2a1e7eba8f033a68a2a44a8/src/main/java/com/williambl/haema/mixin/LivingEntityMixin.java#L46) are two mods which do.

Unfortunately, this requires a redirect mixin - however, this same redirect being needed for multiple mods (thus making them incompatible) suggests that it is appropriate for Fabric API.

CC: @martmists-gh

williambl committed:
Add ServerPlayerEvents.BEFORE_DEATH event

williambl committed:
Fix checkstyle

liach commented on fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java:
```diff
@@ -45,6 +46,21 @@
+		boolean result = true;
+
+		for (BeforeDeath callback : callbacks) {
+			result &= callback.beforeDeath(oldPlayer, damageSource, amount);
```
hmm, does the callback gets called if the result is already false? and should it be called in that case?

williambl replied:
They do, I was thinking about doing it either way but decided to do it this way (i.e. not short-circuiting) just to avoid surprising behaviour ('I registered this callback, why does it not always run?' etc.). If the consensus is that it's better short-circuiting, then I can change it to be done that way.

liach replied:
It indeed does not short circuit, which is imo good enough. :+1:

liach commented:
I have a question: In your example, you give the player 3 health as the player holds apple on death. What if the player wears a helmet and a mod gives the helmet wearer like 5 health? We need a way to address such conflicts.

williambl commented:
> I have a question: In your example, you give the player 3 health as the player holds apple on death. What if the player wears a helmet and a mod gives the helmet wearer like 5 health? We need a way to address such conflicts.

I don't quite understand the issue - surely two mods could do that for basically any event, with the last-running event listener 'winning'?

The 'set to 3 health' thing is only in the example so that you don't die next tick due to health being below zero - there are other ways of preventing next-tick-death, such as giving a status effect, doing player.heal(), or by mixing into and modifying the logic used by isDead - I'd imagine many mods wouldn't be setting health directly.

liach reviewed (approved):
Sure, other mods can just inject in mods' handler methods if they want to make sure they register their listener after

Juuxel commented on fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java:
```diff
@@ -69,6 +85,19 @@
+		 * @param amount the amount of damage that has killed the player
+		 * @return whether or not the player should die
+		 */
+		boolean beforeDeath(ServerPlayerEntity player, DamageSource source, float amount);
```
`damage` would be a nicer and more descriptive name for `amount`

williambl replied:
It seems a bit weird to have `source` but also `damage` - maybe `damageSource` and `damageAmount`?

Technici4n replied:
Yeah why not have `damageSource` and `damageAmount`.

williambl replied:
Done

Martmists-GH commented:
Gunpowder already offers events for this incompatibility, those might be a good point of reference.

Technici4n commented:
If the point of this event is to cancel player deaths, I think it should:
* be called `CANCEL_DEATH`,
* allow returning `true` to cancel death,
* short-circuit subsequent handling if `true` was returned so that death is not cancelled multiple times.

The javadoc should mention this, and also mention that this event is called before totems of undying have a chance to be used (and returning `true` will prevent totems of undying from being used).
I would also prefer `return true if ..., false otherwise` over `return whether ...`.

What happens if a handler cancels the death but doesn't change the player health? Does this leave the player in some kind of "undead" state, and should we guard against it?

williambl commented:
> If the point of this event is to cancel player deaths, I think it should:
> 
>     * be called `CANCEL_DEATH`,
> 
>     * allow returning `true` to cancel death,
> 
>     * short-circuit subsequent handling if `true` was returned so that death is not cancelled multiple times.
> 
> 
> The javadoc should mention this, and also mention that this event is called before totems of undying have a chance to be used (and returning `true` will prevent totems of undying from being used).
> I would also prefer `return true if ..., false otherwise` over `return whether ...`.

Sure, all makes sense to me.

> What happens if a handler cancels the death but doesn't change the player health? Does this leave the player in some kind of "undead" state, and should we guard against it?

Vanilla checks for health<=0 each frame (through isDead), and kills if true; so the player will die next frame. It's assumed that the listener will do something to prevent this, for example:
 - a server-side minigame mod teleporting the player into some dedicated 'respawn room' and setting their health to 20.0 again
 - a mod that redoes death mechanics switching the player over to the mod's own play-mode, where death doesn't apply
 - in the case of Haema, we bypass for vampires the vanilla code that checks for death on tick, so that vampires can survive on 0 health (as long as the fatal damage isn't done by certain damage sources, hence the need for this event)

Technici4n commented:
> > What happens if a handler cancels the death but doesn't change the player health? Does this leave the player in some kind of "undead" state, and should we guard against it?
> 
> Vanilla checks for health<=0 each frame (through isDead), and kills if true; so the player will die next frame. It's assumed that the listener will do something to prevent this, for example:

Ok, thanks! That is also worth adding to the javadoc imo, so someone willing to do something specific on death doesn't have to spend 3 hours reverse-engineering Minecraft code. :smile:

The content is good, I am just making sure that someone who is not very familiar with Minecraft's death handling can get a head start by reading the javadoc of the event. :)

LambdAurora commented:
I think that if we have a `CANCEL` event we should have 2 other events, one which notifies that the death got cancelled and another one which notifies the death.
This is a thing commonly requested for events PR I believe.

Technici4n commented:
> I think that if we have a `CANCEL` event we should have 2 other events, one which notifies that the death got cancelled

How can the handler tell if it did cancel the death or not if there are multiple handlers that could have cancelled the death? Imo needs to be a single event that does both.
> and another one which notifies the death.

Possibly, but this is out of scope for this simple compatibility PR. (Imho)

williambl committed:
Apply changes based on PR feedback
 - Replace BEFORE_DEATH with CANCEL_DEATH
 - return true to cancel, false not to cancel
 - short-circuiting
 - rename params

williambl commented:
Made a few changes in response to feedback - see latest commit message

Technici4n reviewed (approved):
Looks good, thanks! I will test this in a bit.

NinjaPhenix commented on fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java:
```diff
@@ -45,6 +47,30 @@
+	 *     apply</ul>
+	 * </li>
+	 */
+	public static final Event<CancelDeath> CANCEL_DEATH = EventFactory.createArrayBacked(CancelDeath.class, callbacks -> (oldPlayer, damageSource, damageAmount) -> {
+		for (CancelDeath callback : callbacks) {
+			if (callback.cancelDeath(oldPlayer, damageSource, damageAmount)) {
```
Imo I'd change `oldPlayer` to just `player` as that's what is used in the event parameters.

```suggestion
	public static final Event<CancelDeath> CANCEL_DEATH = EventFactory.createArrayBacked(CancelDeath.class, callbacks -> (player, damageSource, damageAmount) -> {
		for (CancelDeath callback : callbacks) {
			if (callback.cancelDeath(player, damageSource, damageAmount)) {
```

williambl replied:
That's what I get for copy-pasting the copyplayerevent :sweat_smile:

NinjaPhenix reviewed (changes requested):
Looks good.

williambl commented on fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/mixin/entity/event/LivingEntityMixin.java:
```diff
@@ -17,24 +17,41 @@
+
+	@Redirect(method = "damage", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/LivingEntity;isDead()Z", ordinal = 1))
+	boolean beforePlayerKilled(LivingEntity livingEntity, DamageSource source, float amount) {
+		if (livingEntity instanceof PlayerEntity) {
+			return isDead() && !ServerPlayerEvents.CANCEL_DEATH.invoker().cancelDeath((ServerPlayerEntity) livingEntity, source, amount);
```
Whoops, didn't catch that. Thanks.

williambl committed:
Replace `PlayerEntity` cast with `ServerPlayerEntity` cast

Co-authored-by: NinjaPhenix <5214513+NinjaPhenix@users.noreply.github.com>

williambl committed:
Rename `oldPlayer` param to `player`

Co-authored-by: NinjaPhenix <5214513+NinjaPhenix@users.noreply.github.com>

ghost commented:
@williambl FYI, PR now fails checkstyle due to extra import and missing blank line, see https://github.com/FabricMC/fabric/pull/1394/checks?check_run_id=2231683199#step:5:1859
or run the checkstyle task I guess.

williambl commented:
Yeah, I'll fix it in the morning, I don't want to deal with checkstyle at 10pm :tiny_potato:

williambl committed:
Fix checkstyle

williambl commented:
Checkstyle is now passing :)

liach commented on fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java:
```diff
@@ -45,6 +47,31 @@
+	 *
+	 * <p>Mods can cancel this to keep the player alive.
+	 *
+	 * <p>Vanilla checks for player health <= 0 each frame (with {@link LivingEntity#isDead()}), and kills if true -
```
```suggestion
	 * <p>Vanilla checks for player health {@code <= 0} each tick (with {@link LivingEntity#isDead()}), and kills if true -
```
This shouldn't be frame. This is server player event, so it should be a tick. If it's a frame as in fps on client, move this to client events.

williambl replied:
Whoops, thanks.

liach commented on fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java:
```diff
@@ -45,6 +47,31 @@
+	 * <p>Mods can cancel this to keep the player alive.
+	 *
+	 * <p>Vanilla checks for player health <= 0 each frame (with {@link LivingEntity#isDead()}), and kills if true -
+	 * so the player will die next frame if this is cancelled. It's assumed that the listener will do something to
```
```suggestion
	 * so the player will die next tick if this is not cancelled. It's assumed that the listener will do something to
```

williambl replied:
Should be 'so the player will still die next tick if this event is cancelled' - I'll commit it when I'm at my computer.

williambl committed:
Update fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java

Co-authored-by: liach <7806504+liach@users.noreply.github.com>

williambl committed:
Improve javadoc

williambl committed:
Merge branch 'player-death-event' of github.com:williambl/fabric into player-death-event

liach commented on fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java:
```diff
@@ -45,6 +47,31 @@
+	 *
+	 * <p>Mods can cancel this to keep the player alive.
+	 *
+	 * <p>Vanilla checks for player health {@code <= 0} each frame (with {@link LivingEntity#isDead()}), and kills if true -
```
```suggestion
	 * <p>Vanilla checks for player health {@code <= 0} each tick (with {@link LivingEntity#isDead()}), and kills if true -
```
Somehow....

williambl replied:
merging skill: 0

williambl committed:
Update javadoc: frame -> tick

Co-authored-by: liach <7806504+liach@users.noreply.github.com>

williambl commented:
A thought: is this redirect really needed? An inject into `tryUseTotem` could be used instead:
![image](https://user-images.githubusercontent.com/5115825/113871883-dfca3800-97a2-11eb-918a-361f55300c9c.png)

williambl commented:
Although, injecting into `tryUseTotem` means that the damage amount cannot be accessed. Doing this would work:
```java
LivingEntityMixin {
    private int damageAmountTmp;
}
```

```java
LivingEntity#damage(source, amount) {
[...]

if (this.isDead()) {
:INJECT_1
    damageAmountTmp = amount;
:END_INJECT_1
    if (!this.tryUseTotem(source)) {

[...]
}
```

```java
LivingEntity#tryUseTotem(source) {
:HEAD_INJECT
cir.setReturnValue(ServerPlayerEvents.CANCEL_DEATH.invoker().cancelDeath((ServerPlayerEntity) livingEntity, source, damageAmountTmp));
:END_HEAD_INJECT
}
```

liach commented:
Hmm, can you capture locals for redirects in mixin? Haven't used mixin in a while so not sure. If yes we probably can pass the damage if it's part of the local variables

williambl commented:
Possibly, but this way of doing it uses no redirects.

liach commented:
Imo we shouldn't design our api because the limits of mixin (such as that on redirects). If we lose too much valuable information if we make an alternative injection, then I'd prefer to still use a redirect.

Also, you can check out the current use case of mods and see how important is the damage value (we already got the damage source). If it's not important you can just drop the damage value info altogether.

modmuss50 commented on fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java:
```diff
@@ -69,6 +96,19 @@
 	}
 
+	@FunctionalInterface
+	public interface CancelDeath {
```
Is this the best name? It can be used for more than just cancel'ing death.

I would have 2 events here, an AllowDeath and a BeforeDeath - this would allow mods to cancel death and for others to lisen to it knowing that the player is actually going to die. See the ScreenMouseEvents for a similar example.

Technici4n replied:
Imo this is out of scope, this PR is just a simple compat hook to allow multiple mods to cancel deaths in non-conflicting ways. I don't see why we need to debate additional events instead of just merging this simple addition.

Technici4n commented:
I think having access to the damage is useful, e.g. if you want to add an armor that only allows kills with a low damage. I think the current iteration of the PR is fine.

i509VCB commented on fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/mixin/entity/event/LivingEntityMixin.java:
```diff
@@ -17,24 +17,40 @@
 		ServerEntityCombatEvents.AFTER_KILLED_OTHER_ENTITY.invoker().afterKilledOtherEntity((ServerWorld) ((LivingEntity) (Object) this).world, attacker, (LivingEntity) (Object) this);
 	}
+
+	@Redirect(method = "damage", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/LivingEntity;isDead()Z", ordinal = 1))
```
I believe there is a better injection for this? I will try to look into it since damage is a weird spot people forget to not call `super.damage` in.

modmuss50 replied:
Quite possibly, I imagine this is quite safe as its just the player.

williambl replied:
> I believe there is a better injection for this? I will try to look into it since damage is a weird spot people forget to not call `super.damage` in.

I expressed this idea in above comments: https://github.com/FabricMC/fabric/pull/1394#issuecomment-814902555 https://github.com/FabricMC/fabric/pull/1394#issuecomment-814908478

Technici4n commented:
![image](https://user-images.githubusercontent.com/13494793/115127235-44706880-9fd5-11eb-891b-bddc41741cc7.png)
Working perfectly fine on dedicated servers (server is the test mod, and the client is vanilla), and I believe the scope, documentation and injection point are alright. Marking this as `merge-me-please`.

Player3324 reviewed (changes requested):
Thanks for the contribution, the changes are generally fine but Tech's naming comment unfortunately missed the conventions we already established in the screen API and a few other places.

Please rename the event field to ALLOW_DEATH, the interface name to AllowDeath and the interface method to allowDeath to be consistent with other APIs. This inverts the sementic meaning of the return value so documentation and the check in the invocation loop need to be adjusted.

Technici4n commented:
Hello @williambl, just wondering if you are going to be fixing this PR, or if someone else should do it?

williambl commented:
Sure. I'm not actually planning on using it myself anymore (I prefer injecting into canUseTotem (especially since I don't need damage amount)), but I suppose it'll be useful for others who want to do this without needing a mixin.

williambl committed:
Apply Player's feedback

williambl commented:
Done.

Technici4n reviewed (approved):
Thanks, the update looks correct!

Player3324 reviewed (approved):
Good now, thanks!

immibis commented:
Suggest rebasing these 11 commits down to 1

i509VCB commented:
Squash at merge time should be fine.
